### DISCUSSION TITLE
Middleware stacking and recursive initialization

### DISCUSSION THREAD

**Discussion by euri10** [encode/starlette#1161] [Q&A] [answered]
long story short I was writing a logging middleware whose handlers are initialized in the `__init__` of the middleware and ended up with multiple logs.

it seems to me the `__init__` of the middlewarea are recursively called  because of the `build_middleware_stack` method and I'm not sure if this is intended or a bug.

here's a test case that shows what I think is an issue:

```
class SimpleInitializableMiddleware:
    def __init__(self, app, counter: int):
        self.app = app
        self.counter = counter
        print(self.counter)


def test_middleware_stack_init():
    app = Starlette()
    app.add_middleware(SimpleInitializableMiddleware, counter=1)
    app.add_middleware(SimpleInitializableMiddleware, counter=2)
    app.add_middleware(SimpleInitializableMiddleware, counter=3)
    app.add_middleware(SimpleInitializableMiddleware, counter=4)
```

I would expect to see `1 2 3 4` as the output.
however it prints `1 1 2 1 2 3 1 2 3 4` because for each add_middlware call it goes down the stack and reinitialize it.

what you think @tomchristie ?

**Comment by thunderamur** [comment]
I think it's a bug. It's wrong to make initialization of the same middleware again and again as many times as other middlewares are added.

**Reply by thunderamur** [reply]
We can use `middleware` kwarg when init `Starlette` instead.
```python3
app = Starlette(middleware=[
    Middleware(SimpleInitializableMiddleware, counter=1),
    Middleware(SimpleInitializableMiddleware, counter=2),
])
```

But `add_middleware` method is incorrect anyway.

**Comment by danilito19** [comment]
Would a solution here be to deprecate `add_middleware` and replace it with a `add_middlewares` method that allows for a list of middlewares to be passed in? That way a user wouldn't have to call `add_middleware` multiple times and is just expected to call the new `add_middlewares` once.

**Reply by danilito19** [reply]
@euri10 if you like this suggestion let me know I"d be happy to create a PR with this.

**Reply by Kludex** [reply]
euri10 is taking some time from OSS projects.

I can answer the question tho. Without confirmation that it's a bug or confirm that the behavior should change, the PR will not be accepted. 

In case it's a bug, a PR is welcome.

In case we confirm that the behavior should change, we should deprecate the current behavior, and at the same time introduce the new one. This can potentially break some projects. - the PR is also welcome here in case the condition is satisfied. :)

**Reply by danilito19** [reply]
As a maintainer are you able to classify it as a bug or a new feature? If you say bug, I can PR. If you say feature, who PRs the deprecation warning, if there is one? I can do that if you let me know.

**Reply by Kludex** [reply]
I'm fine with the current behavior.

Let's read what others have to say first ✌️

**Reply by Kludex** [reply]
Besides, Starlette class has the `middleware` parameter.

**Reply by Kludex** [reply]
I'd be willing to accept a PR mentioning this behavior on the docs tho

**Reply by nliccione** [reply]
I do like the simplicity of adding/creating middleware with FastAPI. However, I do have a few concerns about the current implementation.

There will definitely be a performance hit applications take on startup (i.e the cost of initializing each middleware). For applications that have several middlewares installed this will definitely slow them down on startup which is a concern for me and other applications use case. This is the case for both the decorator `@app.middleware("http")` and `app.add_middleware()` .

Although Starlette does have a middleware parameter, I noticed that not all applications use that strategy for adding middleware. For example, [Netflix dispatch](https://github.com/Netflix/dispatch/blob/master/src/dispatch/main.py#L200). In addition, between that strategy and the other two mentioned above the internals of the behavior isn’t consistent between the 3 strategies for adding/creating middleware.

**Reply by thunderamur** [reply]
Breaking the interface is bad.
I've been thinking about it too, it's the easiest solution. I made 2 options.
```python3
class StarletteFix(Starlette):
    def add_middleware_batch(self, middleware=typing.Sequence[Middleware]) -> None:
        self.user_middleware = middleware + self.user_middleware
        self.middleware_stack = self.build_middleware_stack()
```
```python3
class StarletteFix(Starlette):
    def append_middleware(self, middleware_class: type, **options: typing.Any) -> 'StarletteFix':
        self.user_middleware.insert(0, Middleware(middleware_class, **options))
        return self

    def build_middleware(self) -> None:
        self.middleware_stack = self.build_middleware_stack()
```
But in the end, I think that the best option is not to broke the interface, just describe this moment in the documentation. I suggest adding middleware when creating an application by default in the documentation. The add_middleware() method is shown as optional with an indication of that feature with init of the middlewares.

**Reply by danilito19** [reply]
Breaking an interface is bad, absolutely, but not as bad as the interface making an assumption that no longer holds.  As  @thunderamur himself mentioned 

> It's wrong to make initialization of the same middleware again and again as many times as other middlewares are added.

And while there is one other way to add multiple middlewares, `add_middleware` is still highly visible (if not most visible) and has advantages. This behavior can cause bugs as well as performance issues as described above.

**Reply by thunderamur** [reply]
@danilito19
You're probably right. For myself I decided to use only `Starlette(middleware=...)`

**Reply by thunderamur** [reply]
I only just now read @nliccione's [message](https://github.com/encode/starlette/discussions/1161#discussioncomment-1416255).

Maybe just leave only `Starlette(middleware=...)` as the only one method to add middlewares. And mark `add_middleware()` and `@app.middleware("http")` as deprecated.

**Reply by danilito19** [reply]
@thunderamur I think using only `Starlette(middleware=...)` is a solid suggestion! But I'm hoping you could address a concern I have with only that option.

 I reported this [here](https://github.com/tiangolo/fastapi/issues/3972#issuecomment-932372704) explaining what `MyCustomMiddleware` has to do. Basically, using something akin to `app.add_middleware(s)` is the best for my use case because `MyCustomMiddleware` actually **needs a reference to the original FastAPI app**. 

`Starlette(middleware=...)` hasn't yet created the `app`
`app.add_middleware(s)` has created the app so I can pass it to `MyCustomMiddleware`

Is there a workaround for this? Do you have a suggestion for how to  pass an instance of the FastAPI app when one of the middlewares passed on to `Starlette(middleware=...)` needs it?

```
from starlette.middleware.base import BaseHTTPMiddleware

class MyCustomMiddleware(BaseHTTPMiddleware):
    def __init__(self, app: ASGIApp, original_app: FastAPI = None) -> None:
        self.app = app
        self.original_app = original_app
        self.app_name = self.original_app.title

        self.dispatch_func = self.dispatch_func
       # other critical work that should only happen ONCE ever

    async def dispatch_func(self, request: Request, call_next):
        ....
        # work that relies on calling self.original_app.routes
```

**Reply by thunderamur** [reply]
```python3
    def __init__(self, app: ASGIApp, original_app: FastAPI = None) -> None:
        if hasattr(original_app, 'MyCustomMiddleware'):
            return
        app.MyCustomMiddleware = True
```
I don't check it and I don't like this, nor does the binding to original_app. Binding to original_app looks strange as for me, I would try to not use it.

**Reply by lovelydinosaur** [reply]
> I think using only `Starlette(middleware=...)` is a solid suggestion!

Indeed. Personally I'd prefer to eventually see them properly deprecated, and then later removed.

See... https://github.com/encode/starlette/blob/3282d2364119c205adcd13b4e1db8a53af2202c7/starlette/applications.py#L114-L115


> Basically, using something akin to app.add_middleware(s) is the best for my use case because MyCustomMiddleware actually needs a reference to the original FastAPI app.

I would try to avoid that. If you've got a dependancy like that it's almost certainly avoidable.

Also, the app instance is available during runtime in the scope... https://github.com/encode/starlette/blob/3282d2364119c205adcd13b4e1db8a53af2202c7/starlette/applications.py#L111

**Reply by danilito19** [reply]
I see what you mean. So there will be no way for my middleware to have access to the `app` and its router during my middleware's initialization.

**Reply by lovelydinosaur** [reply]
No. Although I'd query what it *actually* needs. Presumably it need access to some information *on* the app, rather than the app itself.

**Reply by danilito19** [reply]
As mentioned above, the middleware needs the app's router registration to compute a list of routes.

**Reply by adriangb** [reply]
@danilito19 would https://github.com/encode/starlette/pull/1649 help?

**Reply by danilito19** [reply]
@adriangb I think so? The main request was that any middleware created with Starlette would have access to the original app and thus the app's router. From my brief understanding of this PR, it looks like that's sort of the case? Can you confirm

**Reply by adriangb** [reply]
What that PR does is let you insert middleware _after_ routing. If your use case was introspecting into the router so that the middleware is only applied to only some routes, then #1649 solves that by letting you apply that middleware to only those routes. If you want a reference to the app, this will _incidentally_ help you, but not necessarily by design.

**Comment by Kludex** [comment, marked as answer]
I've created https://github.com/encode/starlette/issues/2002 from this discussion.
